### PR TITLE
chore(test): clean up between each test run

### PR DESCRIPTION
**Pull request by balazsorban44**
### What?

Run the `beforeEach` and `afterEach` test hooks between each test run, when using the `nextTestSetup` helper.

### Why?

To ensure that test runs are properly isolated and that they don't rely on each other's artifacts.

### How?

For an easier migration, I also added a `DO_NOT_USE_skipTestsIsolation` option to the `nextTestSetup` config. This should be used when we explicitly want to allow tests to depend on each other's artifacts. Note, this should be used very cautiously, and relying on it is likely a sign that the test needs to be rewritten.

If the `skipTestsIsolation: true` option is not set, The `beforeEach` and `afterEach` hooks will automatically reset the Next.js instance between each test.

Closes NEXT-3002

### CODE REVIEW

**Comment by balazsorban44** [comment]
* **#64071** <a href="https://app.graphite.dev/github/pr/vercel/next.js/64071?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @balazsorban44 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
## Failing test suites 
Commit: ecb0d26bbaaa37b7965f1245c334f1623e25764a

`pnpm test-dev test/development/app-dir/app-routes-error/index.test.ts` (PPR)
- app-dir - app routes errors > bad lowercase exports > should print an error when using lowercase "get" in dev
- app-dir - app routes errors > bad lowercase exports > should print an error when using lowercase "head" in dev
- app-dir - app routes errors > bad lowercase exports > should print an error when using lowercase "options" in dev
- app-dir - app routes errors > bad lowercase exports > should print an error when using lowercase "post" in dev
- app-dir - app routes errors > bad lowercase exports > should print an error when using lowercase "put" in dev
- app-dir - app routes errors > bad lowercase exports > should print an error when using lowercase "delete" in dev
- app-dir - app routes errors > bad lowercase exports > should print an error when using lowercase "patch" in dev

<details>
<summary>Expand output</summary>

  ● app-dir - app routes errors › bad lowercase exports › should print an error when using lowercase "get" in dev

    FetchError: request to http://localhost:43319/lowercase/get failed, reason: connect ECONNREFUSED 127.0.0.1:43319

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● app-dir - app routes errors › bad lowercase exports › should print an error when using lowercase "head" in dev

    FetchError: request to http://localhost:43319/lowercase/head failed, reason: connect ECONNREFUSED 127.0.0.1:43319

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● app-dir - app routes errors › bad lowercase exports › should print an error when using lowercase "options" in dev

    FetchError: request to http://localhost:43319/lowercase/options failed, reason: connect ECONNREFUSED 127.0.0.1:43319

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● app-dir - app routes errors › bad lowercase exports › should print an error when using lowercase "post" in dev

    FetchError: request to http://localhost:43319/lowercase/post failed, reason: connect ECONNREFUSED 127.0.0.1:43319

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● app-dir - app routes errors › bad lowercase exports › should print an error when using lowercase "put" in dev

    FetchError: request to http://localhost:43319/lowercase/put failed, reason: connect ECONNREFUSED 127.0.0.1:43319

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● app-dir - app routes errors › bad lowercase exports › should print an error when using lowercase "delete" in dev

    FetchError: request to http://localhost:43319/lowercase/delete failed, reason: connect ECONNREFUSED 127.0.0.1:43319

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● app-dir - app routes errors › bad lowercase exports › should print an error when using lowercase "patch" in dev

    FetchError: request to http://localhost:43319/lowercase/patch failed, reason: connect ECONNREFUSED 127.0.0.1:43319

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

`pnpm test-dev test/development/app-render-error-log/app-render-error-log.test.ts` (PPR)
- app-render-error-log > should log the correct values on app-render error
- app-render-error-log > should log the correct values on app-render error with edge runtime

<details>
<summary>Expand output</summary>

  ● app-render-error-log › should log the correct values on app-render error

    FetchError: request to http://localhost:33279/ failed, reason: connect ECONNREFUSED 127.0.0.1:33279

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

  ● app-render-error-log › should log the correct values on app-render error with edge runtime

    FetchError: request to http://localhost:33279/edge failed, reason: connect ECONNREFUSED 127.0.0.1:33279

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7/node_modules/node-fetch/lib/index.js:1491:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| buildDuration | 13.7s | 13.9s | ⚠️ +159ms |
| buildDurationCached | 7.5s | 6.2s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ✓ |
| nextStartRea..uration (ms) | 403ms | 403ms | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.4 kB | 31.4 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | ✓ |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 99.3 kB | 99.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| index.html gzip | 529 B | 528 B | N/A |
| link.html gzip | 542 B | 541 B | N/A |
| withRouter.html gzip | 525 B | 523 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.4 kB | 95.4 kB | N/A |
| page.js gzip | 3.06 kB | 3.06 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 626 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 97 kB | 97 kB | ✓ |
| app-page-tur..prod.js gzip | 98.8 kB | 98.8 kB | ✓ |
| app-page-tur..prod.js gzip | 93 kB | 93 kB | ✓ |
| app-page.run...dev.js gzip | 144 kB | 144 kB | ✓ |
| app-page.run..prod.js gzip | 91.5 kB | 91.5 kB | ✓ |
| app-route-ex...dev.js gzip | 21.4 kB | 21.4 kB | ✓ |
| app-route-ex..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.1 kB | 21.1 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim...dev.js gzip | 23.1 kB | 23.1 kB | ✓ |
| pages.runtim..prod.js gzip | 22.4 kB | 22.4 kB | ✓ |
| server.runti..prod.js gzip | 51 kB | 51 kB | ✓ |
| Overall change | 945 kB | 945 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js test/clean-up-before-each | Change |
| - | - | - | - |
| 0.pack gzip | 1.58 MB | 1.58 MB | N/A |
| index.pack gzip | 106 kB | 106 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: ecb0d26bbaaa37b7965f1245c334f1623e25764a

File: test/lib/e2e-utils.ts [hunk 1]
```diff
@@ -231,6 +231,13 @@ export async function createNext(
 export function nextTestSetup(
   options: Parameters<typeof createNext>[0] & {
     skipDeployment?: boolean
+    /**
+     * If set to `true`, the Next.js instance won't be cleaned up
+     * between multiple tests in the same suite.
+     *
+     * @note Use with caution: Tests should almost never depend on each other's artifacts.
```
Comment (eps1lon) [inline test/lib/e2e-utils.ts:238]: What would be a legit case? I'm inclined to rename this option to sound more dangerous and make it clear that we want to migrate away.

File: test/lib/e2e-utils.ts [hunk 2]
```diff
@@ -253,18 +260,33 @@ export function nextTestSetup(
...
     beforeAll(async () => {
       next = await createNext(options)
     })
+
+    beforeEach(async () => {
+      if (options.skipStart || options.skipTestsIsolation) return
+      try {
+        await next?.start()
+      } catch (error) {
+        if (error.message === 'next already started') return
```
Comment (eps1lon) [inline test/lib/e2e-utils.ts:274]: When would we have already started in isolated tests?

File: test/lib/e2e-utils.ts (hunk 1 above)
Comment (balazsorban44) [reply test/lib/e2e-utils.ts:238]: Nothing really, maybe, performance. Some tests run on the same output. Maybe their `expect`s could be consolidated under a single `test` though.

**Pull request closed without merging**